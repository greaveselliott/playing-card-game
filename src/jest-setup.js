// Make Enzyme functions available in all test files without importing
import { shallow, render, mount } from 'enzyme';
import React from 'react';

global.shallow = shallow;
global.render = render;
global.mount = mount;

// Fail tests on any warning
console.error = message => {
    throw new Error(message);
};

/**
 * Function to mock an Enzyme component.
 * @param enzymeRender {function} render function.
 * @param Component {JSXComponent} component type to mock.
 * @param componentProps {Object} optional component properties.
 * @param reduxStoreContext {Object} optional redux store context.
 * @returns Enzyme Component.
 */
global.mockComponent = function (enzymeRender, Component, componentProps = {}, reduxStoreContext = {}) {

    let jsxElement = <Component {...componentProps} />;
    const enzymeComponent = enzymeRender(jsxElement, reduxStoreContext);

    return {
        componentProps,
        enzymeComponent
    }
};

/**
 * Function to create data driven tests against poly component configurations
 * @param Component {JSXComponent} component type to mock.
 * @param configurations {Array of objects} An object containing the test configuration for each test case. Each object within this array must have a 'description' and 'props' property.
 * @returns nothing
 */
global.assertComponentProperties = function (Component, configurations) {

    const expectedProperties = ['description', 'props'];

    configurations.forEach( (testCase) => {

        expectedProperties.forEach((property) => {
            if(!testCase.hasOwnProperty(property)) {
                throw new Error(`Error: assertComponentProperties 'configuration' property does not have required '${property}' property.`);
            }
        });

        const { enzymeComponent } = mockComponent(shallow, Component, testCase.props);

        it(testCase.description, () => {
            expect(enzymeComponent).toMatchSnapshot();
        });
    });
};

global.mockResponse = function (status, statusText, response) {
    return new window.Response(response, {
        status: status,
        statusText: statusText,
        headers: {
            'Content-type': 'application/json'
        }
    });
};