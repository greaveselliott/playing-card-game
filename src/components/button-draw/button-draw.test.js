import React from 'react';
import ButtonDraw from './button-draw';
import '../../jest-setup';

describe("ButtonDraw component", () => {  

    it("Should render a ButtonDraw component.", () => {
        const props = {
        };

        const { enzymeComponent } = mockComponent(shallow, ButtonDraw, props);   

        expect(enzymeComponent).toMatchSnapshot();
    });
    
    let data_drive_configuration = [
        {
            description: 'Should trigger on_decrement handler.',
            selector: '.m-button-draw__decrement',
            spy: 'on_decrement'
        },
        {
            description: 'Should trigger on_draw handler.',
            selector: '.m-button-draw__button',
            spy: 'on_draw'
        },
        {
            description: 'Should trigger on_increment handler.',
            selector: '.m-button-draw__increment',
            spy: 'on_increment'
        }
    ];
    
    const spy = {
        [data_drive_configuration[0].spy]: jest.fn(), 
        [data_drive_configuration[1].spy]: jest.fn(), 
        [data_drive_configuration[2].spy]: jest.fn()
    }

    const props = {
        cards_drawing: 1, 
        class_name: "mock_css_class", 
        on_decrement: spy.on_decrement, 
        on_draw: spy.on_draw, 
        on_increment: spy.on_increment
    };

    const { enzymeComponent } = mockComponent(mount, ButtonDraw, props);

    data_drive_configuration.forEach((unit_test) => {
        it(unit_test.description, () => {
            enzymeComponent.find(unit_test.selector).simulate('click');
            expect(spy[unit_test.spy]).toHaveBeenCalled();
        });
    });
});