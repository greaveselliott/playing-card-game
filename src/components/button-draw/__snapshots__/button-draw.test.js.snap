// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDraw component Should render a ButtonDraw component. 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="m-button-draw "
>
    <Button
        class_name="m-button-draw__decrement"
        click_handler={undefined}
        text="-"
    />
    <Button
        class_name="m-button-draw__button"
        click_handler={undefined}
        text="Draw undefined cards"
    />
    <Button
        class_name="m-button-draw__increment"
        click_handler={undefined}
        text="+"
    />
</div>,
  "nodes": Array [
    <div
      className="m-button-draw "
>
      <Button
            class_name="m-button-draw__decrement"
            click_handler={undefined}
            text="-"
      />
      <Button
            class_name="m-button-draw__button"
            click_handler={undefined}
            text="Draw undefined cards"
      />
      <Button
            class_name="m-button-draw__increment"
            click_handler={undefined}
            text="+"
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ButtonDraw />,
      "_debugID": 10,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="m-button-draw "
>
          <Button
                    class_name="m-button-draw__decrement"
                    click_handler={undefined}
                    text="-"
          />
          <Button
                    class_name="m-button-draw__button"
                    click_handler={undefined}
                    text="Draw undefined cards"
          />
          <Button
                    class_name="m-button-draw__increment"
                    click_handler={undefined}
                    text="+"
          />
</div>,
        "_debugID": 11,
        "_renderedOutput": <div
          className="m-button-draw "
>
          <Button
                    class_name="m-button-draw__decrement"
                    click_handler={undefined}
                    text="-"
          />
          <Button
                    class_name="m-button-draw__button"
                    click_handler={undefined}
                    text="Draw undefined cards"
          />
          <Button
                    class_name="m-button-draw__increment"
                    click_handler={undefined}
                    text="+"
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ButtonDraw />,
}
`;
